### DISCUSSION TITLE
Why MicroPython gets flashed to different addresses on different ESP families?

### DISCUSSION THREAD

**Discussion by aivarannamaa** [micropython/micropython#11098] [ESP32]
I'm refactoring an esptool GUI meant for installing MicroPython and CircuitPython and I want to get a better understanding of the address parameter of esptool.

Do I understand correctly that MicroPython uses `0x0` for **all** ESP8266 / ESP32-C3 boards and `0x1000` for **all** ESP32 / ESP32-S2 / ESP32-S3 boards? 

It looks like CircuitPython uses `0x0` for all ESP families they support.

What are the reasons behind the choice of the start address?

Is it likely that a future MicroPython version (or a fork for specific board) starts using different start address for some ESP families compared to current approach?

**Comment by DavesCodeMusings** [comment]
I have an ESP32-S3 board that requires 0x0 as the address. So I don't think it's safe to say **all** boards besides 8266 and C3 use 0x1000.

Here are the addresses for the ones I've tried. All are DevKits from Espressif.

* ESP32-C3 0x0
* ESP32-S3 0x0
* ESP32 0x1000

These addresses are consistent with the flashing instructions given on the micropython.org download site, but like you I'm searching for the reason some use 0x0 while others use 0x1000.

**Reply by aivarannamaa** [reply]
> I have an ESP32-S3 board that requires 0x0 as the address.

Thanks! I don't know why I thought ESP32-S3 goes with 0x1000. This explains why I didn't get my Unexpected Maker's S3Feather working :)

**Reply by DavesCodeMusings** [reply]
I'm still trying to figure it out myself. Besides reading the official documentation for flashing instructions, I can't find a pattern. I thought maybe it was devices with smaller flash sizes (like 8266 and ESP32-C3) using 0x0 and bigger flash (ESP32, S2, and S3) using 0x1000. But the ESP32-S3 using 0x0 disproves that.

If you find a way to determine it programmatically, please post the solution here and I will do the same. Like you, I have been working on a [GUI wrapper for flashing ESPs](https://marketplace.visualstudio.com/items?itemName=DavesCodeMusings.esptool). Right now, I have it configured to prompt the user for the address of 0x0 or 0x1000. It works, but I'd rather have the selection happen automatically... and accurately!

On a side note: as I was troubleshooting "Invalid header" messages from using the wrong address, I gave in and used Thonny to flash the S3. Watching the output window, I saw it chose 0x1000 as well, which didn't work. So it would seem everyone is a bit confused about the ESP32-S3's requirements for the address parameter.
